AutoRec base 작성

kwonskdev opened the pull request:
AutoRec base 작성 (fit/predict/_train_one_epoch 작성)
data 추가 (movielens 100k)
pytest 작성 (movielkens 100k가 autorec에서 잘 돌아가는 지 확인)

kwonskdev committed:
docs: autorec.py docstring 추가

- Encoder, Decoder, AutoRec class의 docstring 추가
- pre-commit-config 파일에서 super-init-not-called를 disable에 추가

kwonskdev committed:
refactor: 컨벤션에 맞게 수정

- 컨벤션에  수정
              no production code change
              no production code change

kwonskdev committed:
<refactor>: 컨벤션에 맞게 수정

- 컨벤션에  수정
              no production code change
              no production code change

kwonskdev committed:
feat: AutoRec의 fit/predict 함수 작성

- AutoRec의 fit/predict 함수 작성
- _train_one_epoch 작성
- 생성자에 loss랑 optimizer 추가
- 기존에 부족했던 docstring 완성

kwonskdev committed:
data: MovieLens 100k 데이터 추가

- MovieLens 100k 데이터 (user_id, item_id) dataframe으로 변환
              no production code change
              no production code change

kwonskdev committed:
feat: 모델의 학습/추천에 사용할 데이터셋 클래스 작성

- 가장 기본적인 형태의 SampleDataset클래스만 작성
              no production code change
              no production code change

kwonskdev committed:
test: model test하는 pytest 함수 작성

- movielens 100k에 대해 autorec 결과 테스트하는 pytest 함수 작성
- test 결과 에러나는 autorec.py, data.py 오타 수정
              no production code change
              no production code change

kwonskdev committed:
merge: Remote autorec-base pull

- Merge branch 'autorec-base' of https://github.com/kwonskdev/deeprecs into autorec-base

kwonskdev commented on deeprecs/tests/model_test.py:
```diff
@@ -0,0 +1,27 @@
+    model : BaseRecommender
+        추천 모델
+    """
+    model = model(input_dim=1683, hidden_dim=128)
```
현재는 autorec만 있어서 parameter를 autorec 기준으로만 작성했는데, 나중에 다른 추천모델들까지 고려하면,
- overload를 쓰거나
- 생성자를 맞추거나
- **kwargs 형태로 작성한다음, valid_arguments만 뽑아오도록 해야 될 것 같습니다.

otzslayer replied:
몇 가지 궁금한 사항이 있는데요.
- 굳이 Lazy fixture를 사용한 이유가 있나요? `pytest.fixture`로 충분하지 않은가요?
- Lazy fixture의 사용 방법이 저게 맞나요?

kwonskdev replied:
몇 가지 궁금한 사항이 있는데요.

- 굳이 Lazy fixture를 사용한 이유가 있나요? pytest.fixture로 충분하지 않은가요?
  - 그냥 fixture의 경우 def test_def(fixture)의 형태로 사용할 때는 문제가 없지만, parametrize에 fixture 사용시 에러가 납니다 (사용한 코드처럼 'autorec'이라는 fixture를 string으로 넘겨야 되는데, fixture로 인지하지 않고 string으로 인지하기 때문에 lazy한 fixture 사용이 필요함). 그래서 lazy_fixture를 사용한거고, 다른 글들 보면 indirect 인자 조정해서 가능하다고 하는데 저는 안되길래 그냥 lazy fixture 사용했습니다. 
- Lazy fixture의 사용 방법이 저게 맞나요?
  - 다른 사용 방법을 생각하시는게 있는건가요? 공식 홈페이지 https://pypi.org/project/pytest-lazy-fixture/에 있는걸 참고했고, 버전차이로 pytest.lazy_fixture가 아닌 pytest_lazyfixture.lazy_fixture로 사용했습니다.

otzslayer replied:
아아 Lazy fixture 사용 방법은 잘 모르지만 string으로 넘기는 부분이 이해가 안됐어서요.
답변 보니까 이해 되네요.
`indirect` 써서 되는지는 나중에 한 번 확인 해봐야겠네요.

kwonskdev commented:
pytest 부분 보시고 문제 없으면 pre-commit에 추가하겠습니다.

가장 상위의 deeprecs 폴더에서
`python -m pytest deeprecs/tests` 실행하시면 됩니다.

otzslayer commented on deeprecs/data/data.py:
```diff
@@ -0,0 +1,34 @@
+        모델의 출력부분
+    """
+
+    def __init__(self, X: torch.Tensor, y: torch.Tensor):
```
AutoEncoder에 `y`도 있나요? 불필요한 메모리 사용이 될 것 같아요.

kwonskdev replied:
나중에 asymmetric autoencoder할 때 필요할 거 같구요, data.py의 class는 자체는 autoencoder 용이라기 보단 범용적으로 만든 class입니다.

otzslayer commented on deeprecs/models/autorec.py:
```diff
@@ -1,77 +1,320 @@
+        batch_size : int
+            한 번 학습할 때 사용할 데이터의 크기
+        """
+        self.train()
```
저는 training mode 켜는걸 `_train_one_epoch()`에서 했었는데 여기서 해야 한 번만 켜게 할 수 있겠네요. 👍

otzslayer commented:
@kwonskdev @yrc726 @HeainLee 
문득 든 생각인데 Getter, Setter를 두는건 모델 내 파라미터에 접근하지 못하게 하려는거잖아요?
접근하는게 큰 문제가 안될 것 같은데 다른 분들 생각은 어떠신가요?

otzslayer commented:
아 그리고 테스트 관련 코드 작성해주셔서 감사합니다.
하나 고민인건 데이터를 이렇게 올려놓고 쓸 것이냐인데 생각 좀 해보겠습니다.

otzslayer commented on deeprecs/models/autorec.py:
```diff
@@ -1,77 +1,320 @@
 
-    def __repr__(self):
-        return self._encoder
+class AutoEncoder(BaseRecommender):
```
컨벤션은 나중에 개선한다고 하면 추가로 필요한 로직은 없어보이네요. 근데 클래스 이름이 `AutoEncoder`인 이유를 알 수 있을까요?

kwonskdev replied:
`AutoRec`이 아닌 `AutoEncoder`인 이유 말씀하시는거죠? 별다른 이유는 없고, Stacked, Denoised 등등 다른 autoencoder 추가할 거 생각해서 그랬습니다.  
지금 드는 생각이지만 AutoRec은 AutoEncoder를 사용하는 추천시스템에 논문 저자가 이름을 붙여준거고, 알고리즘 자체는 autoencoder니까 이게 맞지 않을까 싶긴 합니다. MultiVAERec이 아닌것 처럼요.

otzslayer commented on deeprecs/models/autorec.py:
```diff
@@ -1,77 +1,320 @@
+        """
+        self.train()
+        for _ in range(epochs):
+            train_loader = DataLoader(trainset, batch_size=batch_size)
```
학습용 `DataLoader`는 `shuffle=True`, `num_workers`은 양의 정수로 세팅해야 할 것 같아요.
`num_workers`는 `fit` 메서드에 인자로 받을 수도 있을 것 같고, `shuffle=True`는 반드시 해주시구요.
데이터의 feeding order가 성능에 영향을 미칠 수도 있으니까요.
그리고 나중에는 `verbose` 기준으로 이것저것 출력해주도록 하면 되겠네요.

kwonskdev replied:
`shuffle`과 `num_workers` 둘 다 사용자가 선택하도록 맞는거 같은데(`shuffle=True`를 default로), 그걸 fit의 parameter로 주는게 맞는지는 다른 자료들도 찾아보겠습니다.

otzslayer replied:
제가 알기론 학습할 때는 `shuffle=True`로 두고 추론 시에는 `shuffle=False`로 두는걸로 알고 있습니다.

kwonskdev replied:
생각해보니 선임님 말씀이 맞네요
`shuffle=True` 안하면 epoch이 바뀌어도 계속 같은 batch로 학습하겠네요.
`num_workers` fit 메소드에 인자로 추가하는 거까지 두개 다음 커밋떄 추가하겟습니다.

otzslayer commented on deeprecs/models/autorec.py:
```diff
@@ -1,77 +1,320 @@
+        _Loss
+            AutoRec의 손실함수
+        """
+        return self._loss
```
다른건 모르겠는데 Loss에 대한 Getter/Setter는 있으면 좋을 것 같네요.
Loss는 모델의 아웃풋이니까 뒤에 `_`를 하나 붙여서요.

```python
@property
def loss_(self):
    return self._loss
```

kwonskdev replied:
sklearn 방식대로 하자는 말씀이신거죠? 컨벤션 같은 경우는 나중에 컨벤션이 저희끼리 어느정도 정리되고 나면 한 번에 수정해도 괜찮을까요?

otzslayer replied:
요거는 sklearn 방식 아니고 원래 Python convention 일걸요?
이것도 결국 컨벤션이니까 나중에 싹 한 번에 하시죠.

kwonskdev replied:
sklearn 공식 문서에서는 못찾겠네요...

> sklearn trailing underscore
> https://github.com/rasbt/python-machine-learning-book/blob/master/faq/underscore-convention.md
> The short answer is, the trailing underscore (self.gamma_) in class attributes is a scikit-learn convention to denote "estimated" or "fitted" attributes. The leading underscores are (_sigmoid(self, z)) denote private methods that the user should not bother with.

> sklearn trailing underscore
> https://stackoverflow.com/questions/49138692/what-does-it-mean-for-an-attribute-name-to-end-in-an-underscore
> The convention is used for attributes of estimators that have a meaningful value after fit() was called.

파이썬 pep8에서 single trailing은 파이썬 내부 키워드랑 겹칠 때 사용하라고 돼있습니다
> https://pep8.org/#descriptive-naming-styles
> If your public attribute name collides with a reserved keyword, append a single trailing underscore to your attribute name.

kwonskdev commented:
> @kwonskdev @yrc726 @HeainLee 문득 든 생각인데 Getter, Setter를 두는건 모델 내 파라미터에 접근하지 못하게 하려는거잖아요? 접근하는게 큰 문제가 안될 것 같은데 다른 분들 생각은 어떠신가요?

접근하지 못하게 하는 것도 있지만, 예를 들어서
```
ae = AutoEncoder(input_dim=1, hidden_dim=1)
ae.encoder = 3
```
이런식으로 encoder에 예상(nn.Layer)하지 못한 인자를 주는 경우를 방지하기 위해서도 쓰는게 나을 것 같습니다.

otzslayer commented:
> > @kwonskdev @yrc726 @HeainLee 문득 든 생각인데 Getter, Setter를 두는건 모델 내 파라미터에 접근하지 못하게 하려는거잖아요? 접근하는게 큰 문제가 안될 것 같은데 다른 분들 생각은 어떠신가요?
> 
> 접근하지 못하게 하는 것도 있지만, 예를 들어서
> 
> ```
> ae = AutoEncoder(input_dim=1, hidden_dim=1)
> ae.encoder = 3
> ```
> 
> 이런식으로 encoder에 예상(nn.Layer)하지 못한 인자를 주는 경우를 방지하기 위해서도 쓰는게 나을 것 같습니다.

나중에 써야하는 것과 안써도 되는 것들은 좀 구분해야겠네요 ㅋㅋㅋ

otzslayer reviewed (approved):
일단 approve합니다. 자잘한 부분은 나중에 수정하시죠.

kwonskdev commented:
> > > @kwonskdev @yrc726 @HeainLee 문득 든 생각인데 Getter, Setter를 두는건 모델 내 파라미터에 접근하지 못하게 하려는거잖아요? 접근하는게 큰 문제가 안될 것 같은데 다른 분들 생각은 어떠신가요?
> > 
> > 
> > 접근하지 못하게 하는 것도 있지만, 예를 들어서
> > ```
> > ae = AutoEncoder(input_dim=1, hidden_dim=1)
> > ae.encoder = 3
> > ```
> > 이런식으로 encoder에 예상(nn.Layer)하지 못한 인자를 주는 경우를 방지하기 위해서도 쓰는게 나을 것 같습니다.
> 
> 나중에 써야하는 것과 안써도 되는 것들은 좀 구분해야겠네요 ㅋㅋㅋ

다른 라이브러리들이 모든 attributes를 setter/getter로 관리하지 않는거 때문에 말씀하시는건가요? 그건 저도 이유가 궁금하네요,,, 어떤건 쓰고 어떤건 안쓰는건지

kwonskdev commented:
> 일단 approve합니다. 자잘한 부분은 나중에 수정하시죠.

넵 알겠습니다! 저희 지금 reviewer가 2명 이상 승인하면 머지되는건가여?

kwonskdev committed:
feat: Change from dataset to dataloader

- When `fit` & `predict`, use `DataLoader` instead of `Dataset`.
- Change dataset class name cuz each recom class use theirs.
- `AEDataset` remove `y` when `__init__`.
- `AutoEncoder` `predict` use `torch.no_grad`.
              no production code change
              no production code change
